### ISSUE TITLE
ACA-py is out-of-sync with ledger after failed schema/credDef write

### ISSUE THREAD

**Issue by quinndevos** [hyperledger/aries-cloudagent-python#2256]
Note: The below steps in _italics_ indicate that no user action is necessary for that step.

Steps to reproduce the problem:
1. If the TAA is not accepted,
2. call `POST /schemas` or `POST /credential-definitions`, then
3. _ACA-py stores the schema/credDef in its database, then_
4. _ACA-py attempts to write the schema/credDef to the ledger, then_
5. _the ledger returns an error because the TAA is not accepted, then_
6. accept the TAA (call `POST /ledger/taa/accept`), then
7. call `POST /schemas` or `POST /credential-definitions` again with the same input as used in step 2, then
8. _ACA-py returns an error to say that the schema/credDef already exists, before even attempting to write to the ledger._

The problem:
- After the TAA is accepted in this scenario, that schema/credDef can no longer be written to the ledger because it is stored in ACA-py's database, so ACA-py incorrectly thinks it has been written to the ledger when in reality it is not on the ledger.

**Comment by swcurran** [comment]
Have ACA-Py deal with this should be OK, but it does get a bit complicated. First, it adds overhead to check if the CredDef has been written to the ledger vs. assuming it has. Granted, that should only affect the rare error situation where you call `POST` multiple times for the same CredDef, which should only happen in scenarios like you describe above.

The bigger concern is what happens if you change the parameters to POST /credential-definitions — e.g. change the size (or existence) of the RevReg.  Now, you have to delete the existing CredDef in the wallet and regenerate it.  Or, ACA-Py might see that the CredDef is in the wallet, try to write to the ledger.  As a user, do you expect the first parameters you sent to `POST` to have been used, or the second?

We could complicate the API and add a way to say “It didn’t work the first time, so resend it” — without allowing you to specify the other parameters.

Bottom line is that I think the answer here is “don’t do that”.  The complexity added is more than the benefit of the situation prevented.

I do recognize that there might be other reasons that writing a CredDef could fail, and with those you would be stuck, without some painful workaround.

I’m going to close, but happy to have feedback to suggest we reopen.  Especially if you think this is a high priority.

**Comment by WadeBarnes** [comment]
The same ledger out of sync issue can occur when `ACAPY_READ_ONLY_LEDGER` is inadvertently set to `true` and you try to write a transaction, such as a CredDef to the ledger.  ACA-Py writes the records to its secure storage and then attempts to write the transactions to the ledger.  The ledger write fails because of the `ACAPY_READ_ONLY_LEDGER` setting, leaving the instance's secure storage and the ledger out of sync with no easy way to recover other than to rewrite the records with different versions or tags.

IMO ACA-Py should not be attempting to write transactions to either it's secure storage or the ledger under conditions that are guaranteed to fail.  The instance is aware of it's own settings and should be able to quickly determine whether a TAA is required and has been accepted.  It should already have this information from the first time it connected to the ledger.  It also knows at startup whether or not `ACAPY_READ_ONLY_LEDGER` has been set.

Perhaps a better way to handle the issue is to ensure write operations are atomic so you have some guarantee that the records are written to both secure storage and the ledger or not at all.  That could be as simple as deleting the newly written record from the wallet in the case of a write failure to the ledger, whatever the reason (TAA, ACAPY_READ_ONLY_LEDGER, or network issues).

**Comment by swcurran** [comment]
Good idea on the dealing with the errors and leaving the app in a useful state.

**Comment by swcurran** [comment]
This issue should be considered as we update the Admin API in merging AnonCreds RS into ACA-Py.

Heads up on this @dbluhm and @usingtechnology

**Comment by swcurran** [comment]
Duplicate of #3031
